### PR TITLE
Added options in New Themes subdialog for frequency and summary tables.

### LINKED ISSUE #7993: Adding options to gt summary table
To follow on from Issue #6770, I wanted to create a new issue to discuss changes to the table sub-dialogs since the design has changed a lot since we first discussed!

We currently have two sub dialogs for summary tables: `sdgFormatSummaryTable` and `sdgSummaryTables`. We want some consistency with the sub-dialogs for `ggplot2`. This means we still want two sub-dialogs, but we want to mix things up a bit between them!

The `gt` table is split up into different parts:
image

We use these parts to create two sub dialogs, one on content, and one on themes

**Sub Dialog: Format Summary Table**
* Five tabs: Facets, Titles, Rows, Columns, Themes
* Don’t worry about Facets, Rows, or Columns tabs for now. We will get to these another time.

image

image

Under themes will be some pre-defined themes, such as the seven themes in the `gtExtras` package. We will add more on this later!

**Sub-Sub Dialog: Table Themes**
We have just found the `gtExtras` package which seems to have a lot of options in it! This is discussed here: https://jthomasmock.github.io/gtExtras/ so requires a bit more time for discussions! As a general idea there will be a few tabs (like seen in the ggplot2 themes dialog):

We have this `tab_options` function in the `gt` package which has a lot of options in it (see [here](https://gt.rstudio.com/reference/tab_options.html)). These functions, alongside those in the `gtExtras` package can be used to create the themes sub-dialog:

* Table (this can contain the `table.` parameters in the `tab_options` function. Perhaps the `container.` options too?)
* Titles (can be created from the `heading.` parameters in the `tab_options` function)
* Columns (this could contain the `column_labels` parameters in the `tab_options` function in `gt`, as well as relevant utility functions from `gtExtras` - such as `gt_color_rows`)
* Rows (this could contain the `gt_color_rows`, `gt_highlight_rows` functions from `gtExtras`. And perhaps, for more advanced levels, the `row_group.`, `row_striping` parameters from `gt::tab_options`?)
* Stub (`stub.`, `stub_row_group.` parameters in `tab_options`)
* Footnotes (`footnotes.` parameters)
* Sourcenotes (`source_notes.` parameters)

@anastasia-mbithe does this help? Are you able to conceptualise the individual tabs using the parameter values outlined above?

For example, for the `Titles` themes tab, we can see from the `tab_options` function in `gt` that there are the following relevant parameters (parameters which start with `heading.`):

```
  heading.background.color = NULL,
  heading.align = NULL,
  heading.title.font.size = NULL,
  heading.title.font.weight = NULL,
  heading.subtitle.font.size = NULL,
  heading.subtitle.font.weight = NULL,
  heading.padding = NULL,
  heading.padding.horizontal = NULL,
  heading.border.bottom.style = NULL,
  heading.border.bottom.width = NULL,
  heading.border.bottom.color = NULL,
  heading.border.lr.style = NULL,
  heading.border.lr.width = NULL,
  heading.border.lr.color = NULL,
```

From this, we can conceptualise the `heading` tab:

image

From playing in R, and reading the help file then options are:
* Background Colour: options of lots of colours ("R/X11 color names and CSS 3.0 color names"). Perhaps use the same colour options as we use in the ggplot2 sub dialogs
* Alignment has three options: `left`, `right`, `center` (default center)
* For font weight, it can be a numerical value from 1-1000, or one of four options: `"normal", "bold", "lighter", "bolder"`. Which do you think makes more sense? I'm not sure myself!
* etc etc

You'll see a lot of the tab options contain a similar set of options. Could it be worth creating a new ucr to implement this therefore just from looking at the sdgThemes tab for the plot options. What do you think?

### PR DESCRIPTION
**Pull request by anastasia-mbithe**
Fixes #7993

I have added controls, parameters, and functions (set R code) needed for the new Themes sub dialog. However, the subject requires more discussion with @lilyclements and @N-thony.

### CODE REVIEW

**Comment by anastasia-mbithe** [comment]
@rdstern, Yes I had a discussion with @lilyclements early this week and I have already resumed working on this.

**Comment by anastasia-mbithe** [comment]
@rdstern This is almost ready and its already working, so testing can be done. (I have been testing through the summary tables dialog). 
@lilyclements you can check if there are other more features to add, apart from footnotes and source notes parameters which I should be done by Friday hopefully.

**Review by rdstern — COMMENTED** [review]
@anastasia-mbithe this is starting to look great.
a) I tried the built-in themes.  They always set back to the first, namely Dark, each time you re-open the sub-dialogue.  And I think the NYTimes theme always gives the Dark theme.  The others seem to work.
b) The defining your own theme is starting to look very comprehensive.  I am assuming this is still work in progress, at least on some of the names for the rows, etc.
c) So tell me when you want me to look more carefully.
d) And separately to that, the reordering of the summaries still doesn't work.  And there will need to be a tab in the Format table to be able to change the names of the summaries.

**Comment by lilyclements** [comment]
@anastasia-mbithe are you happy to work on the changes that @rdstern has outlined above, or did you want me to review as well before you work on the changes (just let me know whichever suits you best)

**Comment by anastasia-mbithe** [comment]
@rdstern, I checked the ucrReorder issue with @N-thony  and it seems it's not an issue from the control since it passes the summaries in the order defined by the user. The issue is from the Rcode/Function which gives the output in alphabetical order. @lilyclements, can also check to confirm that, and see if it's within our power or the package developer's to fix it.

**Comment by rdstern** [comment]
@anastasia-mbithe I hope the order of the summaries can soon be solved.

I also wonder whether the themes progress should now be merged, and then you are free to start on some of the remaining tabs.

I have also checked on the possibility of changing the names of the summaries.  David says that should be easy to do as the summaries are produced, i.e. before using mmtable2.  That would then be a new tab in the Prepare > Data reshape > Summaries dialogue and could then be added to the tables dialogue.  Perhaps make a new issue of this?  It is not as urgent as adding the other features from gt and gtExtras to the Format Tables dialogue.

**Comment by anastasia-mbithe** [comment]
> @anastasia-mbithe I hope the order of the summaries can soon be solved.
> 
> I also wonder whether the themes progress should now be merged, and then you are free to start on some of the remaining tabs.
> 
> I have also checked on the possibility of changing the names of the summaries. David says that should be easy to do as the summaries are produced, i.e. before using mmtable2. That would then be a new tab in the Prepare > Data reshape > Summaries dialogue and could then be added to the tables dialogue. Perhaps make a new issue of this? It is not as urgent as adding the other features from gt and gtExtras to the Format Tables dialogue.

Yes, this is ready for testing, the themes bit, we've disabled some tabs like the row/columns/stub, etc since they are not in use at the moment. Until we implement the remaining tabs in the format sub dialog.

**Comment by anastasia-mbithe** [comment]
@rdstern, since the Format sub dialog has only the Titles and Themes tabs implemented, and the Titles' has footnotes and source notes too, you can test the themes related to titles, footnotes, and source notes.
On the Themes sub dialog, we will enable columns, rows, and stub tabs after the Facets, Rows, and columns tabs on the Format summary are implemented.
Some of the gtExtras Themes are included here
image


I had a discussion with Lily yesterday evening and she will sort out the summary reordering next week, probably on Thursday.
There's more still to be done on these sub-dialogs.

**Comment by rdstern** [comment]
@anastasia-mbithe thank you for the update.

As you have been working on themes, I wonder about the theme I mentioned above, and also about including the gtsummary package, so we can have that theme?

I alsao wonder what else you are working on now, while waiting for the inputs on the tables from @lilyclements ?  Should you be asking @N-thony for other tasks?  There is always plenty still to do!

**Comment by anastasia-mbithe** [comment]
@lilyclements, did you mean the factors multiple receiver?
image

Yes it should be filled. It had a minor bug of not retaining the variables once you reopen the dialog, but that is fixed in this PR now.

**Comment by lloyddewit** [comment]
@anastasia-mbithe I'm not completely clear about the status of this PR. In your view, is it working correctly and ready for testing? Or do you still wish to make changes?
thanks for clarifying

**Comment by anastasia-mbithe** [comment]
@lloyddewit for now, this is ready for testing.

In summary, the following have been implemented in this pr;
1) Fixed the bug on the Factors receiver
2) On sdgFormat summary:- 
     Changed the tabs to Titles, Facets,Rows,Columns and Themes. For now, Facets, rows and columns have been disabled since they are yet to be implemented (an issue  to do with the functions, Lily is following up with mmtable package developer)
3) On the Themes; Select Themes uses the gtExtra package themes.
        For the Manual Theme; all the tabs have been implemented, but the Columns, rows, and stub have been disabled for now since their corresponding functions in sdgFormat summary are yet to be implemented.
4) Lily has fixed the issue of reordering summaries in pr #8211.

I hope this will make it easier when reviewing.

**Comment by lloyddewit** [comment]
@anastasia-mbithe do you still expect changes to this PR from yourself or others? Or is this PR ready for testing? 
Thanks for clarifying

**Comment by anastasia-mbithe** [comment]
@lloyddewit This is ready for testing.

**Comment by lloyddewit** [comment]
@rdstern please could you test? thanks

**Review by rdstern — APPROVED** [review]
@anastasia-mbithe looks good.  And I really like the parallel between themes for ggplot and those for gt.

File: instat/sdgSummaryThemes.vb [hunk 1]
```diff
@@ -17,7 +17,1080 @@
...
+        ucrInputFootnotesMarks.SetParameter(New RParameter("footnotes.marks"))
+        ucrInputFootnotesMarks.SetItems(dctFootnotesMarks)
+        ucrChkFootnotesMarks.AddParameterPresentCondition(True, "footnotes.marks")
+        ucrChkFootnotesMarks.AddParameterPresentCondition(False, "footnotes.marks", False)
+
+        ucrChkFootnotesMultiline.SetText("Footnotes Multiline")
+        ucrInputFootnotesMultiline.SetParameter(New RParameter("footnotes.marks"))
+        ucrInputFootnotesMultiline.SetItems(dctFootnotesMultiline)
+        ucrChkFootnotesMultiline.AddParameterPresentCondition(True, "footnotes.marks")
+        ucrChkFootnotesMultiline.AddParameterPresentCondition(False, "footnotes.marks", False)
```
Comment (lloyddewit) [inline instat/sdgSummaryThemes.vb:706]: are these the correct settings for multiline?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@anastasia-mbithe Thank you, this PR needed a lot of extra code and it all looks neatly and professionally implemented.
My comments are mainly small aesthetic things.

**Comment by anastasia-mbithe** [comment]
@lloyddewit, I made the changes

**Review by lloyddewit — DISMISSED** [review]

**Comment by lloyddewit** [comment]
@rdstern Please could you retest the multiline (the only code that functionaly changed), we can then merge, thanks

**Review by rdstern — DISMISSED** [review]
@anastasia-mbithe and @lilyclements this all seems to work, but I assume it will need to change now to get rid of mmtable2?

But perhaps not?  When the main tables command is merged, then perhaps the mmtable2 disappears automatically.  I am approving, but could lily also approve before it is merged.

**Comment by rdstern** [comment]
@lloyddewit I checked with @lilyclements and would be happy to see this issue merged.

**Comment by anastasia-mbithe** [comment]
@rdstern and @lloyddewit, This branch had some conflicts from the merging of the other related PR. The conflicts are resolved now, but that has dismissed your reviews/approvals.

**Review by lloyddewit — APPROVED** [review]

**Review by rdstern — APPROVED** [review]

**Pull request merged**